Source/CarthageKit/Project.swift

```diff
@@ -1289,14 +1290,78 @@ internal func frameworksInDirectory(_ directoryURL: URL) -> SignalProducer<URL,
 			}
 			return frameworksInURL.count == 1
 		}.filter { url in
-
-			let packageType: PackageType? = Bundle(url: url)?.packageType
-
+			
+			// Try reading the Info.plist of the Framework
+			let bundle = Bundle(url: url)
+			let packageType: PackageType? = bundle?.packageType
+			
 			switch packageType {
 			case .framework?, .bundle?:
 				return true
 			default:
-				return false
+				// In case no Info.plist exists check the Mach-O fileType
+				guard let executablePath = bundle?
+					.executableURL?
+					.resolvingSymlinksInPath()
+					.path else {
+						
+					return false
+				}
+				
+				// Read the Mach header with `objdump` and parse the output.
+				// The output is composed of one or more sets of lines like the following:
+				//
+				// Mach header
+				//      magic cputype  cpusubtype  caps    filetype ncmds sizeofcmds      flags
+				// 0xfeedfacf 16777223          3  0x00           1     8       1720 0x00002000
+				//
+				// See output structure at:
+				// https://llvm.org/viewvc/llvm-project/llvm/trunk/tools/llvm-objdump/MachODump.cpp?view=markup&pathrev=225383###see%C2%B7line%C2%B72745
+
+				let task = Task("/usr/bin/xcrun", arguments: [
+					"objdump",
+					"-macho",
+					"-private-header",
+					"-non-verbose",
+					executablePath
+					]
+				)
+				let fileType = task.launch(standardInput: nil)
+					.ignoreTaskData()
+					.map { String(data: $0, encoding: .utf8) ?? "" }
+					.filter { !$0.isEmpty }
+					.flatMap(.merge) { (output: String) -> SignalProducer<String, NoError> in
+						output.linesProducer
+					}.filter {
+						// Keep the lines with Mach-O magic number indefiers
+						$0.contains(String(format: "%x", MH_CIGAM))
+							|| $0.contains(String(format: "%x", MH_MAGIC))
+							|| $0.contains(String(format: "%x", MH_CIGAM_64))
+							|| $0.contains(String(format: "%x", MH_MAGIC_64))
+					}
+					.take(last: 1)
```

Comment:
So, on the unstated assertion here (‘every `filetype` on a magic number containing line will be the same, so just grab any of them — as implemented: the last one’) — I said earlier that I’d post an implementation where a `reduce` makes this more explicit:

```swift
let acceptedFiletypesAsSpecifiedByMachHeader: Set<Int32> = [
	MH_OBJECT, // Carthage accepts static libraries
	MH_BUNDLE, // Bundles https://github.com/ResearchKit/ResearchKit/blob/1.3.0/ResearchKit/Info.plist#L15-L16
	MH_DYLIB, // or dynamic shared libraries
]

let øøø: Bool = task.launch(standardInput: nil)
	.ignoreTaskData()
	.map { String(data: $0, encoding: .utf8) ?? "" }
	.filter { !$0.isEmpty }
	.flatMap(.merge) { (output: String) -> SignalProducer<(String, String), NoError> in
		output.linesProducer.combinePrevious()
	}
	.filterMap {
		guard
			$0.0.hasPrefix("      magic cputype cpusubtype  caps    filetype ncmds"),
			[MH_CIGAM, MH_MAGIC, MH_CIGAM_64, MH_MAGIC_64]
				.lazy.map({ " " + String(format: "%x", $0) })
				.first(where: $0.1.hasPrefix) != nil
		else { return nil }

		let components = $0.1
			.components(separatedBy: CharacterSet.whitespaces)
			.filter { !$0.isEmpty }

		precondition(components.count == 8, "∆∆∆")

		return Int32(components[4])
	}
	.filter(acceptedFiletypesAsSpecifiedByMachHeader.contains)
	.reduce(into: Set<Int32>()) { $0.insert($1); return }
	.map { $0.count == 1 }
	.single()?
	.value ?? false
```

What do you think?

Reply:
I like the reduce but I think that this line with is maybe too "whitespace" sensitive?

`$0.0.hasPrefix("      magic cputype cpusubtype  caps    filetype ncmds"),`

Reply:
Whitespace is fixed according to <https://llvm.org/viewvc/llvm-project/llvm/trunk/tools/llvm-objdump/MachODump.cpp?view=markup&pathrev=225383###see·line·2745> — IMO, we’d want a silent non-consideration if that output structure were to change (adding a field).

Reply:
I guess we could strip the whitespace from both the output line and our conditional and compare, that would still catch an added field.

Reply:
I'm going to push a few changes according to @mdiep desire to have a separate function.

I'll try to put the reduce too